modules/client/src/main/java/org/apache/ignite/internal/client/tx/ClientTransaction.java

```diff
@@ -101,23 +105,13 @@ public void rollback() throws TransactionException {
     /** {@inheritDoc} */
     @Override
     public CompletableFuture<Void> rollbackAsync() {
-        setState(STATE_ROLLED_BACK);
-
-        return ch.serviceAsync(ClientOp.TX_ROLLBACK, w -> w.out().packLong(id), r -> null);
-    }
-
-    private void setState(int state) {
-        int oldState = this.state.compareAndExchange(STATE_OPEN, state);
-
-        if (oldState == STATE_OPEN) {
-            return;
+        if (!finishFut.compareAndSet(null, new CompletableFuture<>())) {
+            return completedFuture(null);
         }
 
-        String message = oldState == STATE_COMMITTED
-                ? "Transaction is already committed."
-                : "Transaction is already rolled back.";
+        commitState = false;
 
-        throw new TransactionException(message);
+        return ch.serviceAsync(ClientOp.TX_ROLLBACK, w -> w.out().packLong(id), r -> null);
```

Comment:
If we do not wait rollback, we shouldn't wait it at all (no ever it is a first invocation or not)

Reply:
Here is my variant:
```
return finishFut.updateAndGet(fut -> {
    if (fut != null)
        return fut;

    ch.serviceAsync(ClientOp.TX_ROLLBACK, w -> w.out().packLong(id), r -> null);
    
    return CompletableFuture.completedFuture(null);
});
```

Reply:
fixed